### PR TITLE
Improve performance of Memory<T>.Span property getter

### PR DESCRIPTION
**Pull request by GrabYourPitchforks**
Perf results for accessing the `Memory<T>.Span` property getter given different backing objects:

|                      Method |        Toolchain |        Mean |      Error |     StdDev | Scaled |
|---------------------------- |----------------- |------------:|-----------:|-----------:|-------:|
|              CharFromString | baseline (27008) |  3,360.8 ns |  31.849 ns |  28.233 ns |   1.00 |
|              CharFromString |       w/ changes |  2,946.9 ns |  24.783 ns |  21.969 ns |   0.88 |
|                             |                  |             |            |            |        |
|         CharFromArrayOfChar | baseline (27008) |  9,092.7 ns | 156.120 ns | 146.035 ns |   1.00 |
|         CharFromArrayOfChar |       w/ changes |  3,551.1 ns |  40.715 ns |  36.093 ns |   0.39 |
|                             |                  |             |            |            |        |
| CharFromMemoryManagerOfChar | baseline (27008) |  9,088.1 ns |  83.590 ns |  78.190 ns |   1.00 |
| CharFromMemoryManagerOfChar |       w/ changes |  6,460.6 ns |  36.038 ns |  28.136 ns |   0.71 |
|                             |                  |             |            |            |        |
|         ByteFromArrayOfByte | baseline (27008) |  8,479.8 ns | 219.163 ns | 277.172 ns |   1.00 |
|         ByteFromArrayOfByte |       w/ changes |  3,233.9 ns |  30.808 ns |  27.310 ns |   0.38 |
|                             |                  |             |            |            |        |
|        ByteFromArrayOfSByte | baseline (27008) | 42,603.5 ns | 685.866 ns | 641.559 ns |   1.00 |
|        ByteFromArrayOfSByte |       w/ changes |  3,381.6 ns |  71.574 ns | 155.597 ns |   0.08 |
|                             |                  |             |            |            |        |
| ByteFromMemoryManagerOfByte | baseline (27008) |  8,816.3 ns | 175.853 ns | 164.493 ns |   1.00 |
| ByteFromMemoryManagerOfByte |       w/ changes |  6,569.5 ns |  35.088 ns |  31.105 ns |   0.75 |
|                             |                  |             |            |            |        |
|      GetSpanFromEmptyMemory | baseline (27008) |  1,207.4 ns |  22.214 ns |  19.693 ns |   1.00 |
|      GetSpanFromEmptyMemory |       w/ changes |    887.2 ns |   4.268 ns |   3.784 ns |   0.74 |

Various optimizations include:

* We can rely on the `Memory<T>` ctor and factory methods to ensure that an object of an unexpected type never makes it in to the `_object` backing field, even in the face of a torn struct. This allows us to use unsafe code, bypassing the runtime type checks, but still requires that we perform bounds checks as appropriate.

* Single exit from the _Span_ property getter allows the JIT to optimize stack usage and minimize unnecessary data copying.

* Since we're part of coreclr, we can use deep knowledge of runtime object layout and method table layout to further skip some type checks. This is particularly useful in the case where an array is the backing store.

* The _Length_ property getter is once again just a simple field access with no bitmask logic.

* The _Span_ property getter's code gen size shrinks by around 60%.

Unit tests will not pass until the corresponding corefx change comes online since the unit tests perform private reflection over the `Memory<T>` backing fields.

I realize that this may be contentious, especially considering the fact that this proposes probing into the internals of how objects are represented within the runtime. But coreclr code is uniquely positioned to take advantage of these implementation details since the managed code can evolve in sync with any VM changes.

I'm also open to any recommendations people might have for real-world benchmarks so that we can test whether these changes are actually useful in practice.

### CODE REVIEW

File: src/System.Private.CoreLib/shared/System/Memory.cs [hunk 1]
```diff
@@ -209,9 +206,9 @@ public override string ToString()
         {
             if (typeof(T) == typeof(char))
             {
-                return (_object is string str) ? str.Substring(_index, _length & RemoveFlagsBitMask) : Span.ToString();
+                return (_object is string str) ? str.Substring(_index, _length) : Span.ToString();
```
Comment (GrabYourPitchforks) [inline src/System.Private.CoreLib/shared/System/Memory.cs]: This should read `_index & RemoveFlagsBitMask`. Oversight on my part.

File: src/System.Private.CoreLib/shared/System/ReadOnlyMemory.cs [hunk 2]
```diff
@@ -131,9 +133,9 @@ public override string ToString()
         {
             if (typeof(T) == typeof(char))
             {
-                return (_object is string str) ? str.Substring(_index, _length & RemoveFlagsBitMask) : Span.ToString();
+                return (_object is string str) ? str.Substring(_index, _length) : Span.ToString();
```
Comment (GrabYourPitchforks) [inline src/System.Private.CoreLib/shared/System/ReadOnlyMemory.cs]: Same as above - this should be `_index & RemoveFlagsBitMask`.

File: src/System.Private.CoreLib/shared/System/Memory.cs [hunk 3]
```diff
@@ -194,12 +191,12 @@ internal Memory(object obj, int start, int length)
...
         /// The number of items in the memory.
         /// </summary>
-        public int Length => _length & RemoveFlagsBitMask;
+        public int Length => _length;
 
         /// <summary>
         /// Returns true if Length is 0.
         /// </summary>
-        public bool IsEmpty => (_length & RemoveFlagsBitMask) == 0;
+        public bool IsEmpty => _length == 0;
```
Comment (benaadams) [inline src/System.Private.CoreLib/shared/System/Memory.cs]:
```csharp
_length == 0 ? true : false
```
As the Jit isn't so good at returning the results of a test https://github.com/dotnet/coreclr/issues/12306

File: src/System.Private.CoreLib/shared/System/ReadOnlyMemory.cs [hunk 4]
```diff
@@ -116,12 +118,12 @@ internal ReadOnlyMemory(object obj, int start, int length)
...
         /// The number of items in the memory.
         /// </summary>
-        public int Length => _length & RemoveFlagsBitMask;
+        public int Length => _length;
 
         /// <summary>
         /// Returns true if Length is 0.
         /// </summary>
-        public bool IsEmpty => (_length & RemoveFlagsBitMask) == 0;
+        public bool IsEmpty => _length == 0;
```
Comment (benaadams) [inline src/System.Private.CoreLib/shared/System/ReadOnlyMemory.cs]: Also here
```csharp
_length == 0 ? true : false
```

File: src/System.Private.CoreLib/shared/System/Memory.cs (hunk 3 above)
Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Memory.cs]: I doubt that a simple test against zero suffers from the problem tracked by #12306.

Comment (benaadams) [reply src/System.Private.CoreLib/shared/System/Memory.cs]: Was same issue for `string.IsNullOrEmpty` https://github.com/dotnet/coreclr/pull/17512#issuecomment-380488429 after the `cmp` it adds
```asm
+    sete     al
+    movzx    rax, al
+    test     eax, eax
```
Then uses the result of the `test` for the `jmp` rather than the result of the `cmp`

Comment (benaadams) [reply src/System.Private.CoreLib/shared/System/Memory.cs]: Ah, you are right, seems to be fine

File: src/System.Private.CoreLib/shared/System/ReadOnlyMemory.cs (hunk 4 above)
Comment (benaadams) [reply src/System.Private.CoreLib/shared/System/ReadOnlyMemory.cs]: nvm

File: src/System.Private.CoreLib/shared/System/Memory.cs (hunk 3 above)
Comment (mikedn) [reply src/System.Private.CoreLib/shared/System/Memory.cs]: Yep, I'm not aware of any issue involving single condition boolean expressions. `string.IsNullOrEmpty` & co. cases have multiple conditions.

Comment (benaadams) [reply src/System.Private.CoreLib/shared/System/Memory.cs]:
```asm
L0004: cmp dword [rcx+0x8], 0x0
L0008: jnz L0022
```

**Comment by GrabYourPitchforks** [comment]
Just pushed a rebase. No other changes in the latest iteration.

**Comment by GrabYourPitchforks** [comment]
The GetHashCode changes in the latest iteration aren't needed since we should never be able to construct a `Memory<T>` with a null backing object and a non-zero _index_ / _length_. I'll revert them.

**Comment by GrabYourPitchforks** [comment]
I have a pending PR in corefx to react to the unit test breaks that will occur when this PR goes through.

File: src/System.Private.CoreLib/shared/System/Memory.cs [hunk 5]
```diff
@@ -247,50 +239,91 @@ public Memory<T> Slice(int start)
...
+                        // Special-case string since it's the most common for ROM<char>.
+
+                        refToReturn = ref Unsafe.As<char, T>(ref Unsafe.As<string>(tmpObject).GetRawStringData());
+                        lengthOfUnderlyingSpan = Unsafe.As<string>(tmpObject).Length;
+                    }
+                    else if (RuntimeHelpers.ObjectHasComponentSize(tmpObject))
+                    {
+                        // We know the object is not null, it's not a string, and it is variable-length. The only
+                        // remaining option is for it to be a T[] (or a U[] which is blittable to T[], like int[]
+                        // and uint[]). Otherwise somebody used private reflection to set this field, and we're not
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Memory.cs]: > Otherwise somebody used private reflection

Or something like a union struct with explicit layout.

File: src/System.Private.CoreLib/shared/System/Memory.cs [hunk 6]
```diff
@@ -327,37 +360,51 @@ public Span<T> Span
...
+            // It's possible that the below logic could result in an AV if the struct
+            // is torn. This is ok since the caller is expecting to use raw pointers,
+            // and we're not required to keep this as safe as the other Span-based APIs.
+
+            object tmpObject = _object;
+            if (tmpObject != null)
             {
-                // Array is already pre-pinned
-                if (_length < 0)
+                if (typeof(T) == typeof(char) && tmpObject is string s)
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Memory.cs]: In the Span getter, you switched the similar code to using GetType and an unsafe cast for getting the string.  Is there a reason this differs / can't do the same thing?  Should the other one be converted back to match this? etc.

**Comment by jkotas** [comment]
I am ok with depending on internals like this as long as it is within CoreLib.

I am a bit worried about subtle GC holes that `GetObjectMethodTablePointer` may introduce in certain situations.

File: src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs [hunk 7]
```diff
@@ -194,6 +196,45 @@ public static bool IsReferenceOrContainsReferences<T>()
...
+        {
+            Debug.Assert(obj != null);
+
+            // We know that the first data field in any managed object is immediately after the
+            // method table pointer, so just back up one pointer and immediately deref.
+            // This is not ideal in terms of minimizing instruction count but is the best we can do at the moment.
+
+            return Unsafe.Add(ref Unsafe.As<byte, IntPtr>(ref JitHelpers.GetPinningHelper(obj).m_data), -1);
+
+            // Ideally this method would be replaced by the VM with:
```
Comment (jkotas) [inline src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs]: I do not think we would ever want this. This would mean hacking the JIT to allow derefencing objectrefs as pointers. It is invalid IL and it should stay as invalid IL.

I think what we would want here is to make methodtable pointer to be a field and fetch the field here, similar to how it is done in CoreRT.

File: src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs [hunk 8]
```diff
@@ -194,6 +196,45 @@ public static bool IsReferenceOrContainsReferences<T>()
             // See getILIntrinsicImplementation for how this happens.
             throw new InvalidOperationException();
         }
+
+        // Returns true iff the object has a component size;
+        // i.e., is variable length like string, array, Utf8String.
```
Comment (jkotas) [inline src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs]: Nit: We do not have `Utf8String` and it is unclear whether we will ever have it. It may be a bit premature to start mentioning it in comments.

**Comment by danmosemsft** [comment]
Cc @tannergooding @eerhardt

Nice to see this as this has come up as a problem in ML.NET when they tried to use Span more.

File: src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs (hunk 8 above)
Comment (GrabYourPitchforks) [reply src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs]: Good eye - this was an oversight on my part. This code was copied from the feature/utf8string branch and I forgot to clean up the comments. Will fix.

**Comment by GrabYourPitchforks** [comment]
@jkotas You mentioned a potential GC hole in `GetObjectMethodTablePointer`. Can you elaborate? I _believe_ this should be GC-safe, since getting the `ref int` to the first field will be an interior managed pointer back into the object, which is GC-tracked. If the object doesn't have any data fields, then the managed pointer should point _just past_ the end of the object, which the GC treats as equivalently to an interior pointer into the object itself. Then backing up by sizeof(void*) should backtrack no earlier than the original object reference pointer, which is still GC-tracked.

One of my assumptions above may be incorrect. If so we can change the method to be more reliable.

File: src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs (hunk 7 above)
Comment (GrabYourPitchforks) [reply src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs]: As an experiment, I tried writing an `[Intrinsic]` method that was VM-replaced by the mentioned IL sequence. The release VM and JIT handled it just fine (and produced the desired codegen), but the checked JIT hit an assert and failed. I'm sure other parts of the JIT would get angry if I tried to force this through. :)

Really all this gets us in the end is that we eliminate a single instruction. It allows us to turn this (the codegen from this PR):

```asm
lea rax, [rbx + 8] ; rbx = obj ref, rax = ref obj._firstField
mov rax, qword ptr [rax - 8] ; rax = pMethodTable
; .. remainder of logic ..
```

Into this (note no `lea` instruction or SIB syntax):

```asm
mov rax, qword ptr [rbx] ; rbx = obj ref, rax = pMethodTable
; .. remainder of logic ..
```

File: src/System.Private.CoreLib/shared/System/Memory.cs (hunk 5 above)
Comment (GrabYourPitchforks) [reply src/System.Private.CoreLib/shared/System/Memory.cs]: Sure, but explicit layout is unsafe by construction.

```cs
  [StructLayout(LayoutKind.Explicit)]
  private ref struct MyStruct {
    [FieldOffset(0)]
    public ReadOnlySpan<char> TheString;

    [FieldOffset(8)]
    public int Length;
  }

  public static void Main()
  {
    MyStruct mys = default;
    mys.TheString = "Hello!";
    
    foreach (char i in mys.TheString) {
      Console.WriteLine(i);
    }

    mys.Length = 500;

    // below code will read well past the end of the original string
    foreach (char i in mys.TheString) {
      Console.WriteLine(i);
    }
  }
```

**Comment by GrabYourPitchforks** [comment]
@AndyAyersMS We should end up with a managed IntPtr& (which happens to point to offset 0 inside some object) rather than a regular IntPtr. I'm under the impression that the GC should track this as it does any other byref, but Jan's making me question my reality now. 🤔

@jkotas Do you think this is a potential reliability issue that might scuttle this PR? I can perform a benchmark wherein we first pin the object, but I was trying my absolute hardest to avoid anything that might stack-spill, including pinning.

File: src/System.Private.CoreLib/shared/System/Memory.cs (hunk 5 above)
Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/Memory.cs]: Yup.  I was just pointing out it's not limited specifically to private reflection.

**Comment by jkotas** [comment]
> I'm under the impression that the GC should track this as it does any other byref

You are gettting lucky because of we have syncblock index in front of each object. It makes the system to associate `byref` pointing at offset zero with he current object as side-effect (most of the time at least - modulo corner case bugs :-). If we did not have syncblocks in front of each object, `byref` that points at offset zero would have to be associated with end of previous object per ECMA spec.

> @jkotas Do you think this is a potential reliability issue that might scuttle this PR?

I am just saying that somebody may end up spending several weeks to trace down the reliability issue if there is one. If you do due-diligence to check for the potential bugs here, I do not have problem with this change. It may be a good idea to run some stress flavors on this.

**Comment by GrabYourPitchforks** [comment]
FYI I spoke with Maoni about this offline and she indicated that having an interior pointer to the "zero" offset of an object is valid as far as GC reporting / tracking is concerned. We might still need to stress test this since as Jan suggested it's probably not a well exercised condition.

**Comment by jkotas** [comment]
> Unit tests will not pass until the corresponding corefx change comes online since the unit tests perform

The outdated tests should be disabled in https://github.com/dotnet/coreclr/blob/master/tests/CoreFX/CoreFX.issues.json to make the CI green.

File: src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs (hunk 7 above)
Comment (jkotas) [reply src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs]: > the checked JIT hit an assert and failed. I'm sure other parts of the JIT would get angry if I tried to force this through. :)

Right, it is invalid IL and it should stay invalid IL. The comment that says we would want to ideally replace it by `ldarg.0` + `ldind.i` should be deleted.

File: src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs [hunk 9]
```diff
@@ -194,6 +196,45 @@ public static bool IsReferenceOrContainsReferences<T>()
...
+            //
+            // So in effect this method is the equivalent of
+            // return ((MethodTable*)(*obj))->IsStringOrArray();
+
+            Debug.Assert(obj != null);
+            return *(int*)GetObjectMethodTablePointer(obj) < 0;
+        }
+
+        // Given an object reference, returns its MethodTable* as an IntPtr.
+        //[Intrinsic]
```
Comment (jkotas) [inline src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs]: Nit: Delete the commented out attributes

**Comment by GrabYourPitchforks** [comment]
Rebased PR on top of latest master to pick up Jan's earlier changes re: suppressing failing unit tests.

**Comment by GrabYourPitchforks** [comment]
@dotnet-bot test Windows_NT x64 Checked gcstress0x3
@dotnet-bot test Windows_NT x64 Checked corefx_jitstress1

File: src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs [hunk 10]
```diff
@@ -194,6 +196,45 @@ public static bool IsReferenceOrContainsReferences<T>()
             // See getILIntrinsicImplementation for how this happens.
             throw new InvalidOperationException();
         }
+
+        // Returns true iff the object has a component size;
```
Comment (MarcoRossignoli) [inline src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs]: nit: typos `iif`

Comment (gfoidl) [reply src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs]: No, it's correct here -> https://en.wikipedia.org/wiki/If_and_only_if

Comment (MarcoRossignoli) [reply src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs]: Ah ok, never mind! thank's @gfoidl !

File: src/System.Private.CoreLib/shared/System/Memory.cs [hunk 11]
```diff
@@ -407,18 +454,7 @@ public bool Equals(Memory<T> other)
         [EditorBrowsable(EditorBrowsableState.Never)]
         public override int GetHashCode()
         {
-            return _object != null ? CombineHashCodes(_object.GetHashCode(), _index.GetHashCode(), _length.GetHashCode()) : 0;
+            return (_object != null) ? ReadOnlyMemory<T>.CombineHashCodes(RuntimeHelpers.GetHashCode(_object), _index, _length) : 0;
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Memory.cs]: Why not just use `HashCode.Combine` ?

File: src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs (hunk 8 above)
Comment (jkotas) [reply src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs]: Utf8String still mentioned

File: src/System.Private.CoreLib/shared/System/Memory.cs (hunk 11 above)
Comment (GrabYourPitchforks) [reply src/System.Private.CoreLib/shared/System/Memory.cs]: `HashCode.Combine` is randomized because it pessimistically assumes its inputs are attacker-controlled. The `Memory<T>` object doesn't really qualify as attacker-controlled, so it can use a more performance-oriented routine.

File: src/System.Private.CoreLib/shared/System/Memory.cs (hunk 6 above)
Comment (GrabYourPitchforks) [reply src/System.Private.CoreLib/shared/System/Memory.cs]: The `Unsafe.As` construction has better performance so seemed more appropriate for the `Span` property getter. I didn't make the same change here because the `GCHandle.Alloc` call dominates, and I preferred readability over an insignificant performance gain. But if there's benefit to having the two code paths unified we can definitely do so.

File: src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs (hunk 7 above)
Comment (jkotas) [reply src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs]: The right fix for this should be on the VM side, not in the JIT. To make this work, VM should feed the following instruction to the JIT: `ldarg.0` + `ldfld m_pMethodTable`.

Comment (CarolEidt) [reply src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs]: @jkotas - But would it be incorrect for the JIT to fold this? I know that we can get into trouble if we incorrectly optimize address expressions to produce interim byref results that don't point into an object, but it would seem that folding in this case could never be invalid - perhaps I'm missing something?

Comment (jkotas) [reply src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs]: > But would it be incorrect for the JIT to fold this?

Yes, folding the two instructions in the JIT should be fine too.

File: src/System.Private.CoreLib/shared/System/Memory.cs [hunk 12]
```diff
@@ -209,9 +204,9 @@ public override string ToString()
         {
             if (typeof(T) == typeof(char))
             {
-                return (_object is string str) ? str.Substring(_index, _length & RemoveFlagsBitMask) : Span.ToString();
+                return (_object is string str) ? str.Substring(_index, _length) : Span.ToString();
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Memory.cs]: Not necessary for this PR, but I wonder if there's a real benefit to using Substring here, or whether this could just be:
```C#
if (typeof(T) == typeof(char))
{
    return Span.ToString();
}
```

File: src/System.Private.CoreLib/shared/System/Memory.cs (hunk 6 above)
Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/Memory.cs]: Ok, thanks.

File: src/System.Private.CoreLib/shared/System/Memory.cs (hunk 12 above)
Comment (GrabYourPitchforks) [reply src/System.Private.CoreLib/shared/System/Memory.cs]: `Substring`'s primary benefit is that it will return the original string instance if you pass `(0, Length)` as inputs. We could do something similar:

```cs
if (typeof(T) == typeof(char)) {
  if (_object is string str && ((int index = _index) == 0) && ((int length = _length & RemoveFlagsBitmask) == str.Length)) {
    return str;
  } else {
    // somehow rewrite the below line so it's not duplicating the type / index / length checks we just performed above
    return Span.ToString();
  }
}
```

But we're complicating the call site, so it seems unnecessary.

Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/Memory.cs]: > Substring's primary benefit is that it will return the original string instance if you pass (0, Length) as inputs

Yup, good point.

File: src/System.Private.CoreLib/shared/System/Memory.cs (hunk 11 above)
Comment (GrabYourPitchforks) [reply src/System.Private.CoreLib/shared/System/Memory.cs]: That's a good point, especially since the default implementation is referential equality. We can make them use the slower `HashCode` helpers for now for simplicity, and it leaves the door open for us to make targeted perf improvements to these methods in the future if the need arises.

**Comment by GrabYourPitchforks** [comment]
Ping to @dotnet/jit-contrib, wondering if the stress failures noted above are known failures. They don't seem related to this particular code path. (See comment at https://github.com/dotnet/coreclr/pull/20386#issuecomment-435467613.)

**Comment by GrabYourPitchforks** [comment]
Rebased on top of latest master due to last week's `Memory` / `ReadOnlyMemory` changes causing conflicts. Latest iteration also addresses the remaining comment cleanup, further improves `Span` property getter performance (using the slice trick we committed last week), and changes `GetHashCode` to be simpler but subject to a higher performance hit.

**Comment by GrabYourPitchforks** [comment]
@dotnet-bot test Tizen armel Cross Checked Innerloop Build and Test
@dotnet-bot test Windows_NT x64 Checked CoreFX Tests

**Comment by GrabYourPitchforks** [comment]
Heard from @tannergooding that the stress test is a known failure. The particular failing test case takes so long to run that it hits a timeout when running under GC stress mode.

File: tests/CoreFX/CoreFX.issues.json [hunk 13]
```diff
@@ -486,6 +486,14 @@
                 {
                     "name": "System.Buffers.Text.Tests.FormatterTests.TestFormatterDecimal",
                     "reason": "https://github.com/dotnet/coreclr/pull/19775"
+                },
+                {
+                    "name": "System.SpanTests.MemoryMarshalTests.CreateFromPinnedArrayIntSliceRemainsPinned",
```
Comment (ahsonkhan) [inline tests/CoreFX/CoreFX.issues.json]: @GrabYourPitchforks, can we revert this change now?
